File: JTSArjuna/JTSArjuna-1.2.2/Gandiva/lib/Common/Thread/Semaphore.cc
/*
 * Copyright (C) 1994, 1995, 1996,
 *
 * Department of Computing Science,
 * University of Newcastle upon Tyne,
 * Newcastle upon Tyne,
 * UK.
 *
 * $Id: Semaphore.cc,v 1.1 1998/01/12 13:06:46 nmcl Exp $
 */

#include <os/iostream.h>

#ifndef COMMON_THREAD_SEMAPHORE_H_
#  include <Common/Thread/Semaphore.h>
#endif

ClassName* Semaphore::_className = (ClassName*) 0;

ostream& Semaphore::print (ostream& strm) const
{
    strm << "[ " << uid() << " " << className() << " ]";
    return strm;
}

const ClassName& Semaphore::name ()
{
    if (_className == (ClassName*) 0)
	_className = new ClassName("Semaphore");

    return *_className;
}

const ClassName& Semaphore::className () const
{
    return Semaphore::name();
}

void* Semaphore::castup (const ClassName& _type) const
{
    if (_type == Semaphore::name())
	return (void*) this;
    else
	return (void*) 0;
}

Semaphore* Semaphore::castup (Resource* toCast)
{
    if (toCast == (Resource*) 0)
	return (Semaphore*) 0;
    else
	return (Semaphore*) toCast->castup(Semaphore::name());
}

#ifdef NO_INLINES
#  define SEMAPHORE_CC_
#  include <Common/Thread/Semaphore.n>
#  undef SEMAPHORE_CC_
#endif

File: JavaArjuna/v0.1.0/classes/com/arjuna/JavaArjuna/Semaphore/Semaphore.java
/*
 * Copyright (C) 1996, 1997, 1998,
 *
 * Department of Computing Science,
 * University of Newcastle upon Tyne,
 * Newcastle upon Tyne,
 * UK.
 *
 * $Id: Semaphore.java,v 1.5 1998/07/06 13:26:53 nmcl Exp $
 */

package com.arjuna.JavaArjuna.Semaphore;

import com.arjuna.JavaArjuna.Semaphore.SemaphoreImple;
import com.arjuna.JavaArjuna.Common.*;

import java.io.IOException;

public class Semaphore
{

public static final int SM_LOCKED = 0;
public static final int SM_UNLOCKED = 1;
public static final int SM_WOULD_BLOCK = 2;
public static final int SM_ERROR = 3;

    /**
     * Create a semaphore with the default implementation.
     */
    
public Semaphore (String key)
    {
	/*
	 * In the original JavaArjuna there was only one Semaphore implementation.
	 */

	_imple = new BasicSemaphore(key);
    }

    /*
     * https://github.com/nmcl/sandbox/issues/75
public void finalize ()
    {
	_imple = null;
    }
    */

public int lock ()
    {
	return ((_imple != null) ? _imple.lock() : Semaphore.SM_ERROR);
    }
    
public int unlock ()
    {
	return ((_imple != null) ? _imple.unlock() : Semaphore.SM_ERROR);
    }
    
public int tryLock ()
    {
	return ((_imple != null) ? _imple.tryLock() : Semaphore.SM_ERROR);
    }

public String className ()
    {
	return ((_imple != null) ? _imple.className() : null);
    }
    
private SemaphoreImple _imple;
    
};


